# Phase 3 — SQL Business Analysis

## Objective

Analyze the cleaned Olist Brazilian E-Commerce dataset using SQL to answer key business questions related to:

- Revenue and sales
- Customer behavior
- Delivery performance
- Customer experience
- Seller performance
- Product categories
- Geographic patterns

All SQL results will be validated against the cleaned datasets created during Phase 2.

**Phase 3 Status: IN PROGRESS**

In [1]:
import pandas as pd
import sqlite3
import os

conn = sqlite3.connect("../data/olist_ecommerce.db")

print("SQLite database connection created successfully.")

SQLite database connection created successfully.


In [2]:
# Load cleaned datasets

customers_sql = pd.read_csv("../data/cleaned/customers_clean.csv")
orders_sql = pd.read_csv("../data/cleaned/orders_clean.csv")
order_items_sql = pd.read_csv("../data/cleaned/order_items_clean.csv")
payments_sql = pd.read_csv("../data/cleaned/payments_clean.csv")
reviews_sql = pd.read_csv("../data/cleaned/reviews_clean.csv")
products_sql = pd.read_csv("../data/cleaned/products_clean.csv")
sellers_sql = pd.read_csv("../data/cleaned/sellers_clean.csv")
category_translation_sql = pd.read_csv(
    "../data/cleaned/category_translation_clean.csv"
)

print("All cleaned datasets loaded successfully.")

All cleaned datasets loaded successfully.


In [3]:
customers_sql.to_sql(
    "customers",
    conn,
    if_exists="replace",
    index=False
)

orders_sql.to_sql(
    "orders",
    conn,
    if_exists="replace",
    index=False
)

order_items_sql.to_sql(
    "order_items",
    conn,
    if_exists="replace",
    index=False
)

payments_sql.to_sql(
    "payments",
    conn,
    if_exists="replace",
    index=False
)

reviews_sql.to_sql(
    "reviews",
    conn,
    if_exists="replace",
    index=False
)

products_sql.to_sql(
    "products",
    conn,
    if_exists="replace",
    index=False
)

sellers_sql.to_sql(
    "sellers",
    conn,
    if_exists="replace",
    index=False
)

category_translation_sql.to_sql(
    "category_translation",
    conn,
    if_exists="replace",
    index=False
)

print("All datasets loaded into SQLite successfully.")

All datasets loaded into SQLite successfully.


In [4]:
validation_query = """
SELECT 'customers' AS table_name, COUNT(*) AS row_count FROM customers
UNION ALL
SELECT 'orders', COUNT(*) FROM orders
UNION ALL
SELECT 'order_items', COUNT(*) FROM order_items
UNION ALL
SELECT 'payments', COUNT(*) FROM payments
UNION ALL
SELECT 'reviews', COUNT(*) FROM reviews
UNION ALL
SELECT 'products', COUNT(*) FROM products
UNION ALL
SELECT 'sellers', COUNT(*) FROM sellers
UNION ALL
SELECT 'category_translation', COUNT(*) FROM category_translation;
"""

sql_row_counts = pd.read_sql_query(
    validation_query,
    conn
)

sql_row_counts

,table_name,row_count
0,customers,99441
1,orders,99441
2,order_items,112650
3,payments,103886
4,reviews,99224
5,products,32951
6,sellers,3095
7,category_translation,71


In [5]:
expected_counts = {
    "customers": len(customers_sql),
    "orders": len(orders_sql),
    "order_items": len(order_items_sql),
    "payments": len(payments_sql),
    "reviews": len(reviews_sql),
    "products": len(products_sql),
    "sellers": len(sellers_sql),
    "category_translation": len(category_translation_sql)
}

sql_row_counts["expected_count"] = sql_row_counts["table_name"].map(
    expected_counts
)

sql_row_counts["match"] = (
    sql_row_counts["row_count"]
    == sql_row_counts["expected_count"]
)

sql_row_counts

,table_name,row_count,expected_count,match
0,customers,99441,99441,True
1,orders,99441,99441,True
2,order_items,112650,112650,True
3,payments,103886,103886,True
4,reviews,99224,99224,True
5,products,32951,32951,True
6,sellers,3095,3095,True
7,category_translation,71,71,True


In [6]:
print(
    "All SQL row counts match:",
    sql_row_counts["match"].all()
)

All SQL row counts match: True


In [7]:
monthly_revenue_query = """
SELECT
    strftime('%Y-%m', o.order_purchase_timestamp) AS month,
    ROUND(SUM(oi.price), 2) AS product_revenue,
    COUNT(DISTINCT o.order_id) AS total_orders

FROM orders o

JOIN order_items oi
    ON o.order_id = oi.order_id

WHERE o.order_status = 'delivered'

GROUP BY strftime('%Y-%m', o.order_purchase_timestamp)

ORDER BY month;
"""

monthly_revenue = pd.read_sql_query(
    monthly_revenue_query,
    conn
)

monthly_revenue

,month,product_revenue,total_orders
0,2016-09,134.97,1
1,2016-10,40325.11,265
2,2016-12,10.90,1
3,2017-01,111798.36,750
4,2017-02,234223.40,1653
5,2017-03,359198.85,2546
6,2017-04,340669.68,2303
7,2017-05,489338.25,3546
8,2017-06,421923.37,3135
9,2017-07,481604.52,3872


In [7]:
monthly_revenue_query = """
SELECT
    strftime('%Y-%m', o.order_purchase_timestamp) AS month,
    ROUND(SUM(oi.price), 2) AS product_revenue,
    COUNT(DISTINCT o.order_id) AS total_orders

FROM orders o

JOIN order_items oi
    ON o.order_id = oi.order_id

WHERE o.order_status = 'delivered'

GROUP BY strftime('%Y-%m', o.order_purchase_timestamp)

ORDER BY month;
"""

monthly_revenue = pd.read_sql_query(
    monthly_revenue_query,
    conn
)

monthly_revenue

,month,product_revenue,total_orders
0,2016-09,134.97,1
1,2016-10,40325.11,265
2,2016-12,10.90,1
3,2017-01,111798.36,750
4,2017-02,234223.40,1653
5,2017-03,359198.85,2546
6,2017-04,340669.68,2303
7,2017-05,489338.25,3546
8,2017-06,421923.37,3135
9,2017-07,481604.52,3872


In [7]:
monthly_revenue_query = """
SELECT
    strftime('%Y-%m', o.order_purchase_timestamp) AS month,
    ROUND(SUM(oi.price), 2) AS product_revenue,
    COUNT(DISTINCT o.order_id) AS total_orders

FROM orders o

JOIN order_items oi
    ON o.order_id = oi.order_id

WHERE o.order_status = 'delivered'

GROUP BY strftime('%Y-%m', o.order_purchase_timestamp)

ORDER BY month;
"""

monthly_revenue = pd.read_sql_query(
    monthly_revenue_query,
    conn
)

monthly_revenue

,month,product_revenue,total_orders
0,2016-09,134.97,1
1,2016-10,40325.11,265
2,2016-12,10.90,1
3,2017-01,111798.36,750
4,2017-02,234223.40,1653
5,2017-03,359198.85,2546
6,2017-04,340669.68,2303
7,2017-05,489338.25,3546
8,2017-06,421923.37,3135
9,2017-07,481604.52,3872


In [8]:
monthly_revenue_query = """
SELECT
    strftime('%Y-%m', o.order_purchase_timestamp) AS month,

    ROUND(SUM(oi.price), 2) AS product_revenue,

    ROUND(SUM(oi.freight_value), 2) AS freight_revenue,

    ROUND(SUM(oi.price + oi.freight_value), 2) AS total_revenue,

    COUNT(DISTINCT o.order_id) AS total_orders

FROM orders o

JOIN order_items oi
    ON o.order_id = oi.order_id

WHERE o.order_status = 'delivered'

GROUP BY strftime('%Y-%m', o.order_purchase_timestamp)

ORDER BY month;
"""

monthly_revenue = pd.read_sql_query(
    monthly_revenue_query,
    conn
)

monthly_revenue

,month,product_revenue,freight_revenue,total_revenue,total_orders
0,2016-09,134.97,8.49,143.46,1
1,2016-10,40325.11,6165.55,46490.66,265
2,2016-12,10.90,8.72,19.62,1
3,2017-01,111798.36,15684.01,127482.37,750
4,2017-02,234223.40,37015.92,271239.32,1653
5,2017-03,359198.85,55132.10,414330.95,2546
6,2017-04,340669.68,50142.72,390812.40,2303
7,2017-05,489338.25,77513.15,566851.40,3546
8,2017-06,421923.37,68127.00,490050.37,3135
9,2017-07,481604.52,84694.56,566299.08,3872


# 📊 Monthly Product Revenue Analysis

## 🎯 Objective

Analyze the monthly performance of the business by calculating:

- Total Product Revenue
- Total Number of Orders

This helps us understand how sales and customer purchasing activity change over time.

---

## 🗂️ Tables Used

### `orders`
Provides:

- `order_id`
- `order_purchase_timestamp`
- `order_status`

### `order_items`
Provides:

- `order_id`
- `price`

---

## 🔗 Joining Tables

The `orders` and `order_items` tables are connected using:

```sql
JOIN order_items oi
    ON o.order_id = oi.order_id


---

# 📝 MARKDOWN CELL 2 — Monthly Revenue Breakdown

```markdown
# 💰 Monthly Revenue Breakdown Analysis

## 🎯 Objective

Expand the monthly revenue analysis by separating revenue into:

1. Product Revenue
2. Freight Revenue
3. Total Revenue
4. Total Number of Orders

This provides a more complete understanding of the value generated by delivered orders.

---

## 🗂️ Tables Used

### `orders`

Used for:

- `order_id`
- `order_purchase_timestamp`
- `order_status`

### `order_items`

Used for:

- `order_id`
- `price`
- `freight_value`

---

## 💰 Revenue Metrics

### 1. Product Revenue

Calculated using:

```sql
SUM(oi.price)


---

# 📝 MARKDOWN CELL 3 — Average Order Value (AOV)

**Put this Markdown cell BEFORE the AOV query.**

```markdown
# 🛒 Monthly Average Order Value (AOV) Analysis

## 🎯 Objective

Calculate the Average Order Value (AOV) for each month.

Average Order Value helps us understand:

> How much revenue does the business generate on average from each delivered order?

---

## 📐 AOV Formula

The formula for Average Order Value is:

```text
Average Order Value
=
Total Revenue
÷
Total Number of Orders

In [9]:
monthly_aov_query = """
SELECT
    strftime('%Y-%m', o.order_purchase_timestamp) AS month,

    ROUND(
        SUM(oi.price + oi.freight_value) * 1.0 /
        COUNT(DISTINCT o.order_id),
        2
    ) AS average_order_value,

    COUNT(DISTINCT o.order_id) AS total_orders

FROM orders o

JOIN order_items oi
    ON o.order_id = oi.order_id

WHERE o.order_status = 'delivered'

GROUP BY strftime('%Y-%m', o.order_purchase_timestamp)

ORDER BY month;
"""

monthly_aov = pd.read_sql_query(
    monthly_aov_query,
    conn
)

monthly_aov

,month,average_order_value,total_orders
0,2016-09,143.46,1
1,2016-10,175.44,265
2,2016-12,19.62,1
3,2017-01,169.98,750
4,2017-02,164.09,1653
5,2017-03,162.74,2546
6,2017-04,169.70,2303
7,2017-05,159.86,3546
8,2017-06,156.32,3135
9,2017-07,146.25,3872


## 📊 Results and Interpretation

The Average Order Value (AOV) was calculated for each month using:

**AOV = Total Revenue ÷ Number of Unique Orders**

The results show that the Average Order Value generally remained within a relatively stable range throughout the main analysis period.

### Key Observations

- AOV fluctuates across different months rather than continuously increasing.
- Higher total revenue does not necessarily mean customers spend significantly more per order.
- Revenue growth can occur because of:
  1. An increase in the number of orders.
  2. An increase in the average value of each order.
  3. A combination of both.

For example, November 2017 generated the highest total revenue in the analysis. However, its AOV was approximately 158.23.

This indicates that the strong revenue performance during that month was largely associated with a high volume of orders rather than an unusually high Average Order Value.

### Important Note

Months with very few orders, such as September 2016 and December 2016, should be interpreted cautiously because a small number of transactions can produce unstable or unrepresentative AOV values.

Therefore, AOV should be analyzed together with:

- Total Revenue
- Total Number of Orders

This provides a more complete understanding of customer purchasing behavior.

# 📈 Monthly Revenue Growth Analysis

## 🎯 Objective

Analyze how total revenue changes from one month to the next.

This analysis helps identify:

- Revenue growth periods
- Revenue decline periods
- High-growth months
- Changes in business performance over time

---

## 📐 Month-over-Month Growth Formula

Month-over-Month (MoM) growth is calculated using:

**MoM Growth (%) = ((Current Month Revenue - Previous Month Revenue) / Previous Month Revenue) × 100**

A positive value indicates revenue growth.

A negative value indicates a decline in revenue.

---

## 🧠 SQL Concept: Window Functions

A window function performs calculations across related rows without combining them into a single row.

For this analysis, the `LAG()` function is used.

```sql
LAG(total_revenue) OVER (ORDER BY month)

In [10]:
monthly_growth_query = """
WITH monthly_orders AS (
    SELECT
        strftime('%Y-%m', order_purchase_timestamp) AS month,
        COUNT(DISTINCT order_id) AS total_orders
    FROM orders
    WHERE order_status = 'delivered'
    GROUP BY strftime('%Y-%m', order_purchase_timestamp)
)

SELECT
    month,
    total_orders,
    LAG(total_orders) OVER (ORDER BY month) AS previous_month_orders,
    ROUND(
        (
            total_orders - LAG(total_orders) OVER (ORDER BY month)
        ) * 100.0
        / LAG(total_orders) OVER (ORDER BY month),
        2
    ) AS order_growth_percentage
FROM monthly_orders
ORDER BY month;
"""

monthly_growth = pd.read_sql_query(
    monthly_growth_query,
    conn
)

monthly_growth

,month,total_orders,previous_month_orders,order_growth_percentage
0,2016-09,1,NaN,NaN
1,2016-10,265,1.0,26400.00
2,2016-12,1,265.0,-99.62
3,2017-01,750,1.0,74900.00
4,2017-02,1653,750.0,120.40
5,2017-03,2546,1653.0,54.02
6,2017-04,2303,2546.0,-9.54
7,2017-05,3546,2303.0,53.97
8,2017-06,3135,3546.0,-11.59
9,2017-07,3872,3135.0,23.51


# 📈 Monthly Order Growth Analysis

## Objective

Analyze how the number of delivered orders changes from one month to the next.

This helps identify:

- Growth periods
- Declining periods
- Seasonal patterns
- Sudden changes in customer demand

---

## SQL Concepts Used

### 1. CTE (Common Table Expression)

A CTE creates a temporary result set that can be used within the same SQL query.

```sql
WITH monthly_orders AS (
    SELECT ...
)

In [11]:
category_revenue_query = """
SELECT
    COALESCE(
        ct.product_category_name_english,
        p.product_category_name
    ) AS product_category,

    ROUND(SUM(oi.price), 2) AS product_revenue,

    COUNT(DISTINCT o.order_id) AS total_orders,

    ROUND(
        SUM(oi.price) * 1.0 /
        COUNT(DISTINCT o.order_id),
        2
    ) AS revenue_per_order

FROM orders o

JOIN order_items oi
    ON o.order_id = oi.order_id

JOIN products p
    ON oi.product_id = p.product_id

LEFT JOIN category_translation ct
    ON p.product_category_name = ct.product_category_name

WHERE o.order_status = 'delivered'

GROUP BY
    COALESCE(
        ct.product_category_name_english,
        p.product_category_name
    )

ORDER BY product_revenue DESC;
"""

category_revenue = pd.read_sql_query(
    category_revenue_query,
    conn
)

category_revenue.head(15)

,product_category,product_revenue,total_orders,revenue_per_order
0,health_beauty,1233131.72,8647,142.61
1,watches_gifts,1166176.98,5495,212.23
2,bed_bath_table,1023434.76,9272,110.38
3,sports_leisure,954852.55,7530,126.81
4,computers_accessories,888724.61,6530,136.10
5,furniture_decor,711927.69,6307,112.88
6,housewares,615628.69,5743,107.20
7,cool_stuff,610204.10,3559,171.45
8,auto,578966.65,3810,151.96
9,toys,471286.48,3804,123.89


In [12]:
category_revenue_query = """
SELECT
    COALESCE(
        ct.product_category_name_english,
        p.product_category_name
    ) AS product_category,

    ROUND(SUM(oi.price), 2) AS product_revenue,

    COUNT(DISTINCT o.order_id) AS total_orders,

    ROUND(
        SUM(oi.price) * 1.0 /
        COUNT(DISTINCT o.order_id),
        2
    ) AS revenue_per_order

FROM orders o

JOIN order_items oi
    ON o.order_id = oi.order_id

JOIN products p
    ON oi.product_id = p.product_id

LEFT JOIN category_translation ct
    ON p.product_category_name = ct.product_category_name

WHERE o.order_status = 'delivered'

GROUP BY
    COALESCE(
        ct.product_category_name_english,
        p.product_category_name
    )

ORDER BY product_revenue DESC;
"""

category_revenue = pd.read_sql_query(
    category_revenue_query,
    conn
)

category_revenue.head(15)

,product_category,product_revenue,total_orders,revenue_per_order
0,health_beauty,1233131.72,8647,142.61
1,watches_gifts,1166176.98,5495,212.23
2,bed_bath_table,1023434.76,9272,110.38
3,sports_leisure,954852.55,7530,126.81
4,computers_accessories,888724.61,6530,136.10
5,furniture_decor,711927.69,6307,112.88
6,housewares,615628.69,5743,107.20
7,cool_stuff,610204.10,3559,171.45
8,auto,578966.65,3810,151.96
9,toys,471286.48,3804,123.89


# 📦 Product Category Revenue Analysis

## Objective

Identify which product categories generate the highest product revenue.

This analysis helps understand:

- Which categories contribute the most revenue
- Which categories have the highest order volume
- Which categories generate higher product value per order

---

## Tables Used

### orders

Used to:

- Identify orders
- Filter only delivered orders
- Connect orders with order items

### order_items

Used to obtain:

- Product ID
- Product price

The `price` column represents the product price.

### products

Used to identify the category associated with each product.

### category_translation

Used to translate Portuguese product category names into English.

---

## SQL Concepts Used

### 1. INNER JOIN

```sql
JOIN order_items oi
    ON o.order_id = oi.order_id

In [13]:
category_volume_query = """
SELECT
    COALESCE(
        ct.product_category_name_english,
        p.product_category_name
    ) AS product_category,

    COUNT(DISTINCT o.order_id) AS total_orders,

    COUNT(*) AS total_items,

    ROUND(SUM(oi.price), 2) AS product_revenue

FROM orders o

JOIN order_items oi
    ON o.order_id = oi.order_id

JOIN products p
    ON oi.product_id = p.product_id

LEFT JOIN category_translation ct
    ON p.product_category_name = ct.product_category_name

WHERE o.order_status = 'delivered'

GROUP BY
    COALESCE(
        ct.product_category_name_english,
        p.product_category_name
    )

ORDER BY total_orders DESC;
"""

category_volume = pd.read_sql_query(
    category_volume_query,
    conn
)

category_volume.head(15)

,product_category,total_orders,total_items,product_revenue
0,bed_bath_table,9272,10953,1023434.76
1,health_beauty,8647,9465,1233131.72
2,sports_leisure,7530,8431,954852.55
3,computers_accessories,6530,7644,888724.61
4,furniture_decor,6307,8160,711927.69
5,housewares,5743,6795,615628.69
6,watches_gifts,5495,5859,1166176.98
7,telephony,4093,4430,309860.23
8,auto,3810,4140,578966.65
9,toys,3804,4030,471286.48


In [14]:
customer_state_query = """
SELECT
    c.customer_state,
    COUNT(DISTINCT o.order_id) AS total_orders,
    COUNT(DISTINCT o.customer_id) AS total_customers

FROM orders o

JOIN customers c
    ON o.customer_id = c.customer_id

GROUP BY c.customer_state

ORDER BY total_orders DESC;
"""

customer_state_analysis = pd.read_sql_query(
    customer_state_query,
    conn
)

customer_state_analysis

,customer_state,total_orders,total_customers
0,SP,41746,41746
1,RJ,12852,12852
2,MG,11635,11635
3,RS,5466,5466
4,PR,5045,5045
5,SC,3637,3637
6,BA,3380,3380
7,DF,2140,2140
8,ES,2033,2033
9,GO,2020,2020


In [15]:
unique_customers_state_query = """
SELECT
    customer_state,
    COUNT(DISTINCT customer_unique_id) AS total_unique_customers

FROM customers

GROUP BY customer_state

ORDER BY total_unique_customers DESC;
"""

unique_customers_state = pd.read_sql_query(
    unique_customers_state_query,
    conn
)

unique_customers_state

,customer_state,total_unique_customers
0,SP,40302
1,RJ,12384
2,MG,11259
3,RS,5277
4,PR,4882
5,SC,3534
6,BA,3277
7,DF,2075
8,ES,1964
9,GO,1952


In [16]:
revenue_by_state_query = """
SELECT
    c.customer_state,
    ROUND(SUM(oi.price), 2) AS product_revenue,
    COUNT(DISTINCT o.order_id) AS total_orders

FROM orders o

JOIN customers c
    ON o.customer_id = c.customer_id

JOIN order_items oi
    ON o.order_id = oi.order_id

WHERE o.order_status = 'delivered'

GROUP BY c.customer_state

ORDER BY product_revenue DESC;
"""

revenue_by_state = pd.read_sql_query(
    revenue_by_state_query,
    conn
)

revenue_by_state

,customer_state,product_revenue,total_orders
0,SP,5067633.16,40501
1,RJ,1759651.13,12350
2,MG,1552481.83,11354
3,RS,728897.47,5345
4,PR,666063.51,4923
5,SC,507012.13,3546
6,BA,493584.14,3256
7,DF,296498.41,2080
8,GO,282836.70,1957
9,ES,268643.45,1995


## Revenue by Customer State

### Business Question
Which customer states generate the highest product revenue?

### SQL Query

```sql
SELECT
    c.customer_state,
    ROUND(SUM(oi.price), 2) AS product_revenue,
    COUNT(DISTINCT o.order_id) AS total_orders

FROM orders o

JOIN customers c
    ON o.customer_id = c.customer_id

JOIN order_items oi
    ON o.order_id = oi.order_id

WHERE o.order_status = 'delivered'

GROUP BY c.customer_state

ORDER BY product_revenue DESC;

In [18]:
state_aov_query = """
SELECT
    c.customer_state,

    ROUND(
        SUM(oi.price + oi.freight_value)
        / COUNT(DISTINCT o.order_id),
        2
    ) AS average_order_value,

    COUNT(DISTINCT o.order_id) AS total_orders

FROM orders o

JOIN customers c
    ON o.customer_id = c.customer_id

JOIN order_items oi
    ON o.order_id = oi.order_id

WHERE o.order_status = 'delivered'

GROUP BY c.customer_state

ORDER BY average_order_value DESC;
"""

state_aov = pd.read_sql_query(
    state_aov_query,
    conn
)

state_aov

,customer_state,average_order_value,total_orders
0,PB,266.61,517
1,AC,244.69,80
2,AP,240.92,67
3,AL,237.21,397
4,RO,234.43,243
5,PA,224.13,946
6,PI,220.96,476
7,RR,220.48,41
8,TO,219.00,274
9,RN,212.48,474


## Average Order Value (AOV) by Customer State

### Business Question
Which customer states have the highest average spending per order?

### Formula

AOV = Total Revenue / Number of Orders

### SQL Query

SELECT
    c.customer_state,

    ROUND(
        SUM(oi.price + oi.freight_value)
        / COUNT(DISTINCT o.order_id),
        2
    ) AS average_order_value,

    COUNT(DISTINCT o.order_id) AS total_orders

FROM orders o

JOIN customers c
    ON o.customer_id = c.customer_id

JOIN order_items oi
    ON o.order_id = oi.order_id

WHERE o.order_status = 'delivered'

GROUP BY c.customer_state

ORDER BY average_order_value DESC;

### SQL Concepts

#### SUM()
Calculates the total monetary value.

SUM(oi.price + oi.freight_value)

adds product prices and freight charges.

#### COUNT(DISTINCT)
Counts every order only once.

COUNT(DISTINCT o.order_id)

This is important because one order can contain multiple items.

#### GROUP BY
Groups the analysis by customer state.

GROUP BY c.customer_state

#### ORDER BY DESC
Sorts the results from the highest AOV to the lowest.

### Key Insight

A state with high total revenue does not necessarily have a high
Average Order Value.

SP generates the highest total revenue because of its large order volume,
while PB has a higher average spending per order.

### Important Consideration

States with a small number of orders may have unstable or misleading
AOV values. Always consider total order volume along with AOV.

repeat_customers_query = """
SELECT
    customer_unique_id,
    COUNT(DISTINCT o.order_id) AS total_orders

FROM customers c

JOIN orders o
    ON c.customer_id = o.customer_id

GROUP BY customer_unique_id

HAVING COUNT(DISTINCT o.order_id) > 1

ORDER BY total_orders DESC;
"""

repeat_customers = pd.read_sql_query(
    repeat_customers_query,
    conn
)

repeat_customers

In [20]:
repeat_customers_query = """
SELECT
    customer_unique_id,
    COUNT(DISTINCT o.order_id) AS total_orders

FROM customers c

JOIN orders o
    ON c.customer_id = o.customer_id

GROUP BY customer_unique_id

HAVING COUNT(DISTINCT o.order_id) > 1

ORDER BY total_orders DESC;
"""

repeat_customers = pd.read_sql_query(
    repeat_customers_query,
    conn
)

repeat_customers

,customer_unique_id,total_orders
0,8d50f5eadf50201ccdcedfb9e2ac8455,17
1,3e43e6105506432c953e165fb2acf44c,9
2,ca77025e7201e3b30c44b472ff346268,7
3,6469f99c1f9dfae7733b25662e7f1782,7
4,1b6c7548a2a1f9037c1fd3ddfed95f33,7
...,...,...
2992,00a39521eb40f7012db50455bf083460,2
2993,0058f300f57d7b93c477a131a59b36c3,2
2994,004b45ec5c64187465168251cd1c9c2f,2
2995,004288347e5e88a27ded2bb23747066c,2


## Repeat Customer Analysis

### Business Question

Which customers have placed more than one order?

### SQL Query

SELECT
    customer_unique_id,
    COUNT(DISTINCT o.order_id) AS total_orders

FROM customers c

JOIN orders o
    ON c.customer_id = o.customer_id

GROUP BY customer_unique_id

HAVING COUNT(DISTINCT o.order_id) > 1

ORDER BY total_orders DESC;


### SQL Concepts

#### GROUP BY

GROUP BY customer_unique_id

Groups all orders belonging to the same actual customer.

---

#### COUNT(DISTINCT)

COUNT(DISTINCT o.order_id)

Counts the number of unique orders placed by each customer.

DISTINCT ensures that an order is counted only once.

---

#### HAVING

HAVING COUNT(DISTINCT o.order_id) > 1

Filters grouped results.

Only customers with more than one order are included.

---

### WHERE vs HAVING

WHERE:
Filters individual rows before grouping.

HAVING:
Filters groups after aggregation.

Example:

WHERE o.order_status = 'delivered'

Filters individual orders.

HAVING COUNT(o.order_id) > 1

Filters customer groups based on their number of orders.


### Key Insight

The dataset contains 2,997 customers who placed more than one order.

The highest-frequency customer placed 17 orders.

In [21]:
repeat_customer_rate_query = """
WITH customer_orders AS (

    SELECT
        c.customer_unique_id,
        COUNT(DISTINCT o.order_id) AS total_orders

    FROM customers c

    JOIN orders o
        ON c.customer_id = o.customer_id

    GROUP BY c.customer_unique_id
)

SELECT
    COUNT(*) AS total_unique_customers,

    SUM(
        CASE
            WHEN total_orders > 1 THEN 1
            ELSE 0
        END
    ) AS repeat_customers,

    ROUND(
        SUM(
            CASE
                WHEN total_orders > 1 THEN 1
                ELSE 0
            END
        ) * 100.0 / COUNT(*),
        2
    ) AS repeat_customer_rate

FROM customer_orders;
"""

repeat_customer_rate = pd.read_sql_query(
    repeat_customer_rate_query,
    conn
)

repeat_customer_rate

,total_unique_customers,repeat_customers,repeat_customer_rate
0,96096,2997,3.12


## Repeat Customer Rate Analysis

### Business Question

What percentage of unique customers placed more than one order?

### Formula

Repeat Customer Rate =
(Repeat Customers / Total Unique Customers) × 100


### SQL Query

WITH customer_orders AS (

    SELECT
        c.customer_unique_id,
        COUNT(DISTINCT o.order_id) AS total_orders

    FROM customers c

    JOIN orders o
        ON c.customer_id = o.customer_id

    GROUP BY c.customer_unique_id
)

SELECT
    COUNT(*) AS total_unique_customers,

    SUM(
        CASE
            WHEN total_orders > 1 THEN 1
            ELSE 0
        END
    ) AS repeat_customers,

    ROUND(
        SUM(
            CASE
                WHEN total_orders > 1 THEN 1
                ELSE 0
            END
        ) * 100.0 / COUNT(*),
        2
    ) AS repeat_customer_rate

FROM customer_orders;


### SQL Concepts

#### CTE (Common Table Expression)

WITH customer_orders AS (...)

A CTE creates a temporary named result that can be used in the main query.

Here, customer_orders creates a temporary table where each row represents one unique customer and their total number of orders.


#### CASE WHEN

CASE
    WHEN total_orders > 1 THEN 1
    ELSE 0
END

If a customer has placed more than one order, SQL assigns 1.

Otherwise, SQL assigns 0.


#### SUM()

SUM() adds all the values.

Since repeat customers receive a value of 1, adding them gives the total number of repeat customers.


#### COUNT(*)

COUNT(*) counts all rows.

Since the CTE contains one row per unique customer, COUNT(*) gives the total number of unique customers.


### Result

Total Unique Customers: 96,096

Repeat Customers: 2,997

Repeat Customer Rate: 3.12%


### Key Insight

Only 3.12% of unique customers placed more than one order during the period represented in the dataset.

The repeat customer count was validated against the previous HAVING query, which also returned 2,997 repeat customers.

In [23]:
tables_query = """
SELECT name
FROM sqlite_master
WHERE type='table';
"""

pd.read_sql_query(tables_query, conn)

,name
0,customers
1,orders
2,order_items
3,payments
4,reviews
5,products
6,sellers
7,category_translation


In [24]:
payment_type_query = """
SELECT
    payment_type,

    COUNT(*) AS payment_records,

    COUNT(DISTINCT order_id) AS total_orders,

    ROUND(SUM(payment_value), 2) AS total_payment_value

FROM payments

GROUP BY payment_type

ORDER BY total_payment_value DESC;
"""

payment_type_analysis = pd.read_sql_query(
    payment_type_query,
    conn
)

payment_type_analysis

,payment_type,payment_records,total_orders,total_payment_value
0,credit_card,76795,76505,12542084.19
1,boleto,19784,19784,2869361.27
2,voucher,5775,3866,379436.87
3,debit_card,1529,1528,217989.79
4,not_defined,3,3,0.00


## Payment Type Analysis

### Business Question

Which payment methods are most commonly used, and how much payment value does each method process?

### SQL Query

SELECT
    payment_type,

    COUNT(*) AS payment_records,

    COUNT(DISTINCT order_id) AS total_orders,

    ROUND(SUM(payment_value), 2) AS total_payment_value

FROM payments

GROUP BY payment_type

ORDER BY total_payment_value DESC;


### SQL Concepts

#### GROUP BY

GROUP BY payment_type

Groups all payment records according to their payment method.

Examples:

- credit_card
- boleto
- voucher
- debit_card


#### COUNT(*)

COUNT(*)

Counts the total number of payment records.

An order can potentially have multiple payment records, so payment records are not always equal to the number of orders.


#### COUNT(DISTINCT order_id)

COUNT(DISTINCT order_id)

Counts the number of unique orders associated with each payment type.

DISTINCT ensures that the same order is counted only once.


#### SUM()

SUM(payment_value)

Adds all payment values within each payment type.

This measures the total payment value processed through each method.


#### ROUND()

ROUND(SUM(payment_value), 2)

Rounds the total payment value to two decimal places.


### Key Insights

- Credit card is the dominant payment method.
- Boleto is the second most commonly used payment method.
- Voucher has more payment records than unique orders, indicating that some orders have multiple voucher payment entries.
- Debit card usage is relatively low.

### Important Data Insight

Payment records and unique orders should not automatically be considered the same thing.

An order may contain multiple payment records.

In [25]:
installment_query = """
SELECT
    payment_installments,

    COUNT(*) AS payment_records,

    ROUND(SUM(payment_value), 2) AS total_payment_value,

    ROUND(AVG(payment_value), 2) AS average_payment_value

FROM payments

GROUP BY payment_installments

ORDER BY payment_installments;
"""

installment_analysis = pd.read_sql_query(
    installment_query,
    conn
)

installment_analysis

,payment_installments,payment_records,total_payment_value,average_payment_value
0,0,2,188.63,94.31
1,1,52546,5907233.36,112.42
2,2,12413,1579283.03,127.23
3,3,10461,1491103.80,142.54
4,4,7098,1163907.61,163.98
5,5,5239,961174.30,183.47
6,6,3920,822611.81,209.85
7,7,1626,305157.39,187.67
8,8,4268,1313423.34,307.74
9,9,644,131015.92,203.44


## Payment Installment Analysis

### Business Question

How do customers use installment payments, and how does payment value vary by installment count?


### SQL Query

SELECT
    payment_installments,

    COUNT(*) AS payment_records,

    ROUND(SUM(payment_value), 2) AS total_payment_value,

    ROUND(AVG(payment_value), 2) AS average_payment_value

FROM payments

GROUP BY payment_installments

ORDER BY payment_installments;


### SQL Concepts

#### GROUP BY

GROUP BY payment_installments

Groups payment records based on the number of installments.

Each group represents a specific installment count.


#### COUNT(*)

COUNT(*)

Counts the number of payment records in each installment group.


#### SUM()

SUM(payment_value)

Calculates the total payment value associated with each installment count.


#### AVG()

AVG(payment_value)

Calculates the average payment value for each installment group.

Formula:

Average Payment Value =
Total Payment Value / Number of Payment Records


#### ROUND()

ROUND(..., 2)

Rounds monetary values to two decimal places.


#### ORDER BY

ORDER BY payment_installments

Sorts the results according to the number of installments.


### Key Insights

- One installment is the most commonly used payment option.
- Higher installment counts generally tend to be associated with higher average payment values.
- Very high installment counts have very few records, so their averages should be interpreted cautiously.
- Two payment records have zero installments. These should not automatically be removed without investigation.

In [26]:
payment_installment_type_query = """
SELECT
    payment_type,
    payment_installments,

    COUNT(*) AS payment_records,

    ROUND(SUM(payment_value), 2) AS total_payment_value,

    ROUND(AVG(payment_value), 2) AS average_payment_value

FROM payments

GROUP BY
    payment_type,
    payment_installments

ORDER BY
    payment_type,
    payment_installments;
"""

payment_installment_type_analysis = pd.read_sql_query(
    payment_installment_type_query,
    conn
)

payment_installment_type_analysis

,payment_type,payment_installments,payment_records,total_payment_value,average_payment_value
0,boleto,1,19784,2869361.27,145.03
1,credit_card,0,2,188.63,94.31
2,credit_card,1,25455,2440445.43,95.87
3,credit_card,2,12413,1579283.03,127.23
4,credit_card,3,10461,1491103.80,142.54
5,credit_card,4,7098,1163907.61,163.98
6,credit_card,5,5239,961174.30,183.47
7,credit_card,6,3920,822611.81,209.85
8,credit_card,7,1626,305157.39,187.67
9,credit_card,8,4268,1313423.34,307.74


## Payment Type × Installment Analysis

### Business Question

How does installment usage differ across different payment methods?


### SQL Query

SELECT
    payment_type,
    payment_installments,

    COUNT(*) AS payment_records,

    ROUND(SUM(payment_value), 2) AS total_payment_value,

    ROUND(AVG(payment_value), 2) AS average_payment_value

FROM payments

GROUP BY
    payment_type,
    payment_installments

ORDER BY
    payment_type,
    payment_installments;


### SQL Concepts

#### GROUP BY Multiple Columns

GROUP BY
    payment_type,
    payment_installments

SQL creates a separate group for every unique combination of:

- Payment type
- Number of installments

For example:

credit_card + 1 installment

credit_card + 2 installments

boleto + 1 installment


#### COUNT(*)

COUNT(*)

Counts the number of payment records in each payment type and installment combination.


#### SUM()

SUM(payment_value)

Calculates the total payment value for each group.


#### AVG()

AVG(payment_value)

Calculates the average payment value for each group.


### Key Insights

- Credit cards are the only payment method with substantial installment usage.
- Boleto payments use one installment.
- Debit card payments use one installment.
- Voucher payments use one installment.
- Higher credit card installment counts generally tend to be associated with higher average payment values.
- Very high installment counts have small sample sizes and should be interpreted carefully.
- Two credit card payment records have zero installments. These are retained as dataset edge cases.

In [27]:
order_status_query = """
SELECT
    order_status,

    COUNT(*) AS total_orders,

    ROUND(
        COUNT(*) * 100.0 / SUM(COUNT(*)) OVER (),
        2
    ) AS percentage_of_orders

FROM orders

GROUP BY order_status

ORDER BY total_orders DESC;
"""

order_status_analysis = pd.read_sql_query(
    order_status_query,
    conn
)

order_status_analysis

,order_status,total_orders,percentage_of_orders
0,delivered,96478,97.02
1,shipped,1107,1.11
2,canceled,625,0.63
3,unavailable,609,0.61
4,invoiced,314,0.32
5,processing,301,0.30
6,created,5,0.01
7,approved,2,0.00


## Order Status Analysis

### Business Question

What is the distribution of orders across different order statuses?


### SQL Query

SELECT
    order_status,

    COUNT(*) AS total_orders,

    ROUND(
        COUNT(*) * 100.0 / SUM(COUNT(*)) OVER (),
        2
    ) AS percentage_of_orders

FROM orders

GROUP BY order_status

ORDER BY total_orders DESC;


### SQL Concepts

#### GROUP BY

GROUP BY order_status

Groups orders according to their status.


#### COUNT(*)

COUNT(*)

Counts the number of orders in each status.

Since the orders table contains one row per order,
COUNT(*) correctly represents the number of orders.


#### Window Function: OVER()

SUM(COUNT(*)) OVER ()

First:

COUNT(*)

calculates the number of orders in each status group.

Then:

SUM(...) OVER ()

calculates the total across all groups without collapsing
the results into a single row.


#### Percentage Calculation

COUNT(*) * 100.0 / SUM(COUNT(*)) OVER ()

Calculates:

Orders in a specific status
/
Total orders

× 100


#### ORDER BY

ORDER BY total_orders DESC

Sorts statuses from the highest number of orders
to the lowest.


### Key Insights

- Approximately 97% of orders are delivered.
- The dataset is heavily dominated by delivered orders.
- Non-delivered statuses represent a very small proportion of orders.
- Delivery performance analysis should focus primarily on delivered orders because only delivered orders have a complete delivery timeline.

In [28]:
delivery_time_query = """
SELECT
    COUNT(*) AS delivered_orders,

    ROUND(
        AVG(
            julianday(order_delivered_customer_date)
            - julianday(order_purchase_timestamp)
        ),
        2
    ) AS average_delivery_days,

    ROUND(
        MIN(
            julianday(order_delivered_customer_date)
            - julianday(order_purchase_timestamp)
        ),
        2
    ) AS minimum_delivery_days,

    ROUND(
        MAX(
            julianday(order_delivered_customer_date)
            - julianday(order_purchase_timestamp)
        ),
        2
    ) AS maximum_delivery_days

FROM orders

WHERE order_status = 'delivered'

AND order_purchase_timestamp IS NOT NULL
AND order_delivered_customer_date IS NOT NULL;
"""

delivery_time_analysis = pd.read_sql_query(
    delivery_time_query,
    conn
)

delivery_time_analysis

,delivered_orders,average_delivery_days,minimum_delivery_days,maximum_delivery_days
0,96470,12.56,0.53,209.63


## Average Delivery Time Analysis

### Business Question

How long does it take, on average, for a customer to receive a delivered order?


### SQL Query

SELECT
    COUNT(*) AS delivered_orders,

    ROUND(
        AVG(
            julianday(order_delivered_customer_date)
            - julianday(order_purchase_timestamp)
        ),
        2
    ) AS average_delivery_days,

    ROUND(
        MIN(
            julianday(order_delivered_customer_date)
            - julianday(order_purchase_timestamp)
        ),
        2
    ) AS minimum_delivery_days,

    ROUND(
        MAX(
            julianday(order_delivered_customer_date)
            - julianday(order_purchase_timestamp)
        ),
        2
    ) AS maximum_delivery_days

FROM orders

WHERE order_status = 'delivered'

AND order_purchase_timestamp IS NOT NULL
AND order_delivered_customer_date IS NOT NULL;


### SQL Concepts

#### julianday()

SQLite does not provide a standard DATEDIFF() function.

julianday() converts a date/time value into a numeric value
that allows dates to be subtracted.

Example:

julianday(order_delivered_customer_date)
-
julianday(order_purchase_timestamp)

This calculates the number of days between purchase
and customer delivery.


#### AVG()

AVG() calculates the average delivery duration.


#### MIN()

MIN() finds the shortest delivery duration.


#### MAX()

MAX() finds the longest delivery duration.


#### WHERE

WHERE order_status = 'delivered'

Ensures that only delivered orders are included.

Orders that were canceled, shipped, or still processing
should not be included in actual delivery time analysis.


#### IS NOT NULL

Ensures that both required timestamps exist before
calculating the delivery duration.


### Key Insights

- The average delivery time is approximately 12.56 days.
- 96,470 delivered orders had the required timestamps
  for calculating delivery duration.
- A small number of delivered orders were excluded due
  to missing required date values.
- Delivery times vary substantially, with some extreme
  cases taking more than 200 days.

In [29]:
missing_dates_query = """
SELECT
    COUNT(*) AS total_delivered_orders,

    SUM(
        CASE
            WHEN order_purchase_timestamp IS NULL
            THEN 1
            ELSE 0
        END
    ) AS missing_purchase_date,

    SUM(
        CASE
            WHEN order_delivered_customer_date IS NULL
            THEN 1
            ELSE 0
        END
    ) AS missing_customer_delivery_date,

    SUM(
        CASE
            WHEN order_estimated_delivery_date IS NULL
            THEN 1
            ELSE 0
        END
    ) AS missing_estimated_delivery_date

FROM orders

WHERE order_status = 'delivered';
"""

missing_dates = pd.read_sql_query(
    missing_dates_query,
    conn
)

missing_dates

,total_delivered_orders,missing_purchase_date,missing_customer_delivery_date,missing_estimated_delivery_date
0,96478,0,8,0


In [30]:
valid_delivery_records_query = """
SELECT
    COUNT(*) AS total_delivered_orders,

    COUNT(
        CASE
            WHEN order_purchase_timestamp IS NOT NULL
            AND order_delivered_customer_date IS NOT NULL
            THEN 1
        END
    ) AS valid_delivery_time_records,

    COUNT(
        CASE
            WHEN order_delivered_customer_date IS NOT NULL
            AND order_estimated_delivery_date IS NOT NULL
            THEN 1
        END
    ) AS valid_delivery_delay_records

FROM orders

WHERE order_status = 'delivered';
"""

valid_delivery_records = pd.read_sql_query(
    valid_delivery_records_query,
    conn
)

valid_delivery_records

,total_delivered_orders,valid_delivery_time_records,valid_delivery_delay_records
0,96478,96470,96470


In [31]:
delivery_delay_status_query = """
SELECT
    order_status,

    COUNT(*) AS valid_delivery_delay_records

FROM orders

WHERE order_delivered_customer_date IS NOT NULL
AND order_estimated_delivery_date IS NOT NULL

GROUP BY order_status

ORDER BY valid_delivery_delay_records DESC;
"""

delivery_delay_status = pd.read_sql_query(
    delivery_delay_status_query,
    conn
)

delivery_delay_status

,order_status,valid_delivery_delay_records
0,delivered,96470
1,canceled,6


## Python and SQL Delivery Delay Reconciliation

### Validation Question

Why did the Python analysis show 96,476 valid
delivery-delay records while the SQL analysis showed
96,470 valid delivered records?


### SQL Query

SELECT
    order_status,

    COUNT(*) AS valid_delivery_delay_records

FROM orders

WHERE order_delivered_customer_date IS NOT NULL
AND order_estimated_delivery_date IS NOT NULL

GROUP BY order_status

ORDER BY valid_delivery_delay_records DESC;


### Results

- Delivered orders: 96,470
- Canceled orders: 6

Total valid delivery-delay records:

96,470 + 6 = 96,476


### Explanation

The Python analysis calculated delivery delay across the
entire order-level dataset.

It did not restrict the calculation to only delivered orders.

Therefore, the Python result included:

- 96,470 delivered orders
- 6 canceled orders

Both groups had non-null actual and estimated delivery dates.


### Conclusion

There is no discrepancy between Python and SQL results.

The difference occurred because of different analysis scopes.

For delivery performance metrics such as late-delivery rate,
only delivered orders should be included.

In [32]:
late_delivery_query = """
SELECT
    COUNT(*) AS valid_delivered_orders,

    SUM(
        CASE
            WHEN order_delivered_customer_date
                 > order_estimated_delivery_date
            THEN 1
            ELSE 0
        END
    ) AS late_orders,

    SUM(
        CASE
            WHEN order_delivered_customer_date
                 <= order_estimated_delivery_date
            THEN 1
            ELSE 0
        END
    ) AS on_time_orders,

    ROUND(
        100.0 *
        SUM(
            CASE
                WHEN order_delivered_customer_date
                     > order_estimated_delivery_date
                THEN 1
                ELSE 0
            END
        ) / COUNT(*),
        2
    ) AS late_delivery_rate

FROM orders

WHERE order_status = 'delivered'
AND order_delivered_customer_date IS NOT NULL
AND order_estimated_delivery_date IS NOT NULL;
"""

late_delivery = pd.read_sql_query(
    late_delivery_query,
    conn
)

late_delivery

,valid_delivered_orders,late_orders,on_time_orders,late_delivery_rate
0,96470,7826,88644,8.11


In [33]:
late_delivery_julianday_query = """
SELECT
    COUNT(*) AS valid_delivered_orders,

    SUM(
        CASE
            WHEN (
                julianday(order_delivered_customer_date)
                - julianday(order_estimated_delivery_date)
            ) > 0
            THEN 1
            ELSE 0
        END
    ) AS late_orders

FROM orders

WHERE order_status = 'delivered'
AND order_delivered_customer_date IS NOT NULL
AND order_estimated_delivery_date IS NOT NULL;
"""

late_delivery_julianday = pd.read_sql_query(
    late_delivery_julianday_query,
    conn
)

late_delivery_julianday

,valid_delivered_orders,late_orders
0,96470,7826


In [34]:
late_orders_by_status_query = """
SELECT
    order_status,
    
    COUNT(*) AS late_orders

FROM orders

WHERE order_delivered_customer_date IS NOT NULL
AND order_estimated_delivery_date IS NOT NULL

AND (
    julianday(order_delivered_customer_date)
    - julianday(order_estimated_delivery_date)
) > 0

GROUP BY order_status

ORDER BY late_orders DESC;
"""

late_orders_by_status = pd.read_sql_query(
    late_orders_by_status_query,
    conn
)

late_orders_by_status

,order_status,late_orders
0,delivered,7826
1,canceled,1


## 17.2 Late Delivery Rate

### Business Question

What percentage of delivered orders arrived after their
estimated delivery date?


### Definition

An order is considered late when:

Actual Delivery Date > Estimated Delivery Date


Orders delivered on the estimated date are considered
on time.


### SQL Query

SELECT
    COUNT(*) AS valid_delivered_orders,

    SUM(
        CASE
            WHEN order_delivered_customer_date
                 > order_estimated_delivery_date
            THEN 1
            ELSE 0
        END
    ) AS late_orders,

    SUM(
        CASE
            WHEN order_delivered_customer_date
                 <= order_estimated_delivery_date
            THEN 1
            ELSE 0
        END
    ) AS on_time_orders,

    ROUND(
        100.0 *
        SUM(
            CASE
                WHEN order_delivered_customer_date
                     > order_estimated_delivery_date
                THEN 1
                ELSE 0
            END
        ) / COUNT(*),
        2
    ) AS late_delivery_rate

FROM orders

WHERE order_status = 'delivered'
AND order_delivered_customer_date IS NOT NULL
AND order_estimated_delivery_date IS NOT NULL;


### SQL Concepts

#### CASE WHEN

CASE WHEN provides conditional logic in SQL.

Syntax:

CASE
    WHEN condition THEN value
    ELSE value
END

In this analysis:

If actual delivery date > estimated delivery date:
    Late = 1

Otherwise:
    Late = 0


#### SUM(CASE WHEN ...)

SUM(
    CASE
        WHEN condition THEN 1
        ELSE 0
    END
)

Counts the number of rows satisfying the condition.

Late orders receive 1.
On-time orders receive 0.

SUM() adds all the 1s.


#### COUNT(*)

COUNT(*) counts all rows that remain after the WHERE
conditions are applied.

Here, it represents the number of valid delivered orders.


#### ROUND()

ROUND(value, 2)

Rounds a numerical value to 2 decimal places.


#### 100.0

Using 100.0 ensures decimal arithmetic when calculating
the percentage.


### Formula

Late Delivery Rate =

(Late Orders / Valid Delivered Orders) × 100


### Results

- Valid delivered orders: 96,470
- Late orders: 7,826
- On-time orders: 88,644
- Late delivery rate: 8.11%


### Validation

7,826 + 88,644 = 96,470

Late Delivery Rate:

(7,826 / 96,470) × 100 = 8.11%


### Important Data Scope

Only orders with:

- order_status = 'delivered'
- non-null actual delivery date
- non-null estimated delivery date

are included.

Canceled orders are excluded from the delivery performance KPI.


### Python vs SQL Reconciliation

Python originally showed:

- Late orders: 7,827

SQL showed:

- Delivered late orders: 7,826

Investigation found:

- 1 canceled order also had a positive delivery delay.

Therefore:

7,826 delivered late orders
+    1 canceled late order
-------------------------
7,827 total late records

The Python and SQL results are completely consistent.


### Business Insight

Approximately 8.11% of valid delivered orders arrived
after their estimated delivery date.

Approximately 91.89% arrived on or before the estimated
delivery date.


In [35]:
delivery_state_query = """
SELECT
    c.customer_state,

    COUNT(*) AS valid_delivered_orders,

    SUM(
        CASE
            WHEN o.order_delivered_customer_date
                 > o.order_estimated_delivery_date
            THEN 1
            ELSE 0
        END
    ) AS late_orders,

    ROUND(
        100.0 *
        SUM(
            CASE
                WHEN o.order_delivered_customer_date
                     > o.order_estimated_delivery_date
                THEN 1
                ELSE 0
            END
        ) / COUNT(*),
        2
    ) AS late_delivery_rate

FROM orders o

JOIN customers c
    ON o.customer_id = c.customer_id

WHERE o.order_status = 'delivered'
AND o.order_delivered_customer_date IS NOT NULL
AND o.order_estimated_delivery_date IS NOT NULL

GROUP BY c.customer_state

ORDER BY late_delivery_rate DESC;
"""

delivery_by_state = pd.read_sql_query(
    delivery_state_query,
    conn
)

delivery_by_state

,customer_state,valid_delivered_orders,late_orders,late_delivery_rate
0,AL,397,95,23.93
1,MA,717,141,19.67
2,PI,476,76,15.97
3,CE,1279,196,15.32
4,SE,335,51,15.22
5,BA,3256,457,14.04
6,RJ,12350,1664,13.47
7,TO,274,35,12.77
8,PA,946,117,12.37
9,ES,1995,244,12.23


In [36]:
delivery_state_validation_query = """
SELECT
    SUM(
        CASE
            WHEN o.order_delivered_customer_date
                 > o.order_estimated_delivery_date
            THEN 1
            ELSE 0
        END
    ) AS total_late_orders,

    COUNT(*) AS total_valid_delivered_orders

FROM orders o

JOIN customers c
    ON o.customer_id = c.customer_id

WHERE o.order_status = 'delivered'
AND o.order_delivered_customer_date IS NOT NULL
AND o.order_estimated_delivery_date IS NOT NULL;
"""

delivery_state_validation = pd.read_sql_query(
    delivery_state_validation_query,
    conn
)

delivery_state_validation

,total_late_orders,total_valid_delivered_orders
0,7826,96470


In [37]:
first_delivery_base_query = """
SELECT
    c.customer_unique_id,
    o.order_id,
    o.order_purchase_timestamp,
    o.order_delivered_customer_date,
    o.order_estimated_delivery_date,

    ROW_NUMBER() OVER (
        PARTITION BY c.customer_unique_id
        ORDER BY o.order_purchase_timestamp
    ) AS order_number

FROM orders o

JOIN customers c
    ON o.customer_id = c.customer_id

WHERE o.order_status = 'delivered'
AND o.order_delivered_customer_date IS NOT NULL
AND o.order_estimated_delivery_date IS NOT NULL

ORDER BY
    c.customer_unique_id,
    order_number;
"""

first_delivery_base = pd.read_sql_query(
    first_delivery_base_query,
    conn
)

first_delivery_base.head(10)

,customer_unique_id,order_id,order_purchase_timestamp,order_delivered_customer_date,order_estimated_delivery_date,order_number
0,0000366f3b9a7992bf8c76cfdf3221e2,e22acc9c116caa3f2b7121bbb380d08e,2018-05-10 10:56:27,2018-05-16 20:48:37,2018-05-21,1
1,0000b849f77a49e4a4ce2b2a4ca5be3f,3594e05a005ac4d06a72673270ef9ec9,2018-05-07 11:11:27,2018-05-10 18:02:42,2018-05-15,1
2,0000f46a3911fa3c0805444483337064,b33ec3b699337181488304f362a6b734,2017-03-10 21:05:03,2017-04-05 14:38:47,2017-04-07,1
3,0000f6ccb0745a6a4b88665a16c9f078,41272756ecddd9a9ed0180413cc22fb6,2017-10-12 20:29:41,2017-11-01 21:23:05,2017-11-13,1
4,0004aac84e0df4da2b147fca70cf8255,d957021f1127559cd947b62533f484f7,2017-11-14 19:45:42,2017-11-27 23:08:56,2017-12-05,1
5,0004bd2a26a76fe21f786e4fbd80607f,3e470077b690ea3e3d501cffb5e0c499,2018-04-05 19:33:16,2018-04-07 16:12:43,2018-04-19,1
6,00050ab1314c0e55a6ca13cf7181fecf,d0028facea13f508e880202d7097a5a1,2018-04-20 12:57:23,2018-04-27 12:08:59,2018-05-09,1
7,00053a61a98854899e70ed204dd4bafe,44e608f2db00c74a1fe329de44416a4e,2018-02-28 11:15:41,2018-03-16 19:02:51,2018-03-26,1
8,0005e1862207bf6ccc02e4228effd9a0,ae76bef74b97bcb0b3e355e60d9a6f9c,2017-03-04 23:32:12,2017-03-09 08:33:08,2017-04-06,1
9,0005ef4cd20d2893f0d9fbd94d3c0d97,01b330808c5819a6a3cb79b72f0b8288,2018-03-12 15:22:12,2018-05-05 12:38:38,2018-04-04,1


In [38]:
first_delivery_query = """
WITH ranked_orders AS (

    SELECT
        c.customer_unique_id,
        o.order_id,
        o.order_purchase_timestamp,
        o.order_delivered_customer_date,
        o.order_estimated_delivery_date,

        ROW_NUMBER() OVER (
            PARTITION BY c.customer_unique_id
            ORDER BY o.order_purchase_timestamp
        ) AS order_number

    FROM orders o

    JOIN customers c
        ON o.customer_id = c.customer_id

    WHERE o.order_status = 'delivered'
    AND o.order_delivered_customer_date IS NOT NULL
    AND o.order_estimated_delivery_date IS NOT NULL
)

SELECT
    customer_unique_id,
    order_id,
    order_purchase_timestamp,

    CASE
        WHEN order_delivered_customer_date
             > order_estimated_delivery_date
        THEN 'Late'

        ELSE 'On Time'

    END AS first_delivery_experience

FROM ranked_orders

WHERE order_number = 1;
"""

first_delivery = pd.read_sql_query(
    first_delivery_query,
    conn
)

first_delivery.head(10)

,customer_unique_id,order_id,order_purchase_timestamp,first_delivery_experience
0,0000366f3b9a7992bf8c76cfdf3221e2,e22acc9c116caa3f2b7121bbb380d08e,2018-05-10 10:56:27,On Time
1,0000b849f77a49e4a4ce2b2a4ca5be3f,3594e05a005ac4d06a72673270ef9ec9,2018-05-07 11:11:27,On Time
2,0000f46a3911fa3c0805444483337064,b33ec3b699337181488304f362a6b734,2017-03-10 21:05:03,On Time
3,0000f6ccb0745a6a4b88665a16c9f078,41272756ecddd9a9ed0180413cc22fb6,2017-10-12 20:29:41,On Time
4,0004aac84e0df4da2b147fca70cf8255,d957021f1127559cd947b62533f484f7,2017-11-14 19:45:42,On Time
5,0004bd2a26a76fe21f786e4fbd80607f,3e470077b690ea3e3d501cffb5e0c499,2018-04-05 19:33:16,On Time
6,00050ab1314c0e55a6ca13cf7181fecf,d0028facea13f508e880202d7097a5a1,2018-04-20 12:57:23,On Time
7,00053a61a98854899e70ed204dd4bafe,44e608f2db00c74a1fe329de44416a4e,2018-02-28 11:15:41,On Time
8,0005e1862207bf6ccc02e4228effd9a0,ae76bef74b97bcb0b3e355e60d9a6f9c,2017-03-04 23:32:12,On Time
9,0005ef4cd20d2893f0d9fbd94d3c0d97,01b330808c5819a6a3cb79b72f0b8288,2018-03-12 15:22:12,Late


In [39]:
delivery_repeat_base_query = """
WITH ranked_delivered_orders AS (

    SELECT
        c.customer_unique_id,
        o.order_id,
        o.order_purchase_timestamp,
        o.order_delivered_customer_date,
        o.order_estimated_delivery_date,

        ROW_NUMBER() OVER (
            PARTITION BY c.customer_unique_id
            ORDER BY o.order_purchase_timestamp
        ) AS order_number

    FROM orders o

    JOIN customers c
        ON o.customer_id = c.customer_id

    WHERE o.order_status = 'delivered'
    AND o.order_delivered_customer_date IS NOT NULL
    AND o.order_estimated_delivery_date IS NOT NULL
),

customer_order_counts AS (

    SELECT
        c.customer_unique_id,
        COUNT(o.order_id) AS total_orders

    FROM customers c

    JOIN orders o
        ON c.customer_id = o.customer_id

    GROUP BY c.customer_unique_id
)

SELECT
    r.customer_unique_id,

    CASE
        WHEN r.order_delivered_customer_date
             > r.order_estimated_delivery_date
        THEN 'Late'
        ELSE 'On Time'
    END AS first_delivery_experience,

    c.total_orders,

    CASE
        WHEN c.total_orders > 1
        THEN 'Repeat'
        ELSE 'Not Repeat'
    END AS repeat_status

FROM ranked_delivered_orders r

JOIN customer_order_counts c
    ON r.customer_unique_id = c.customer_unique_id

WHERE r.order_number = 1

ORDER BY r.customer_unique_id;
"""

delivery_repeat_base = pd.read_sql_query(
    delivery_repeat_base_query,
    conn
)

delivery_repeat_base.head(10)

,customer_unique_id,first_delivery_experience,total_orders,repeat_status
0,0000366f3b9a7992bf8c76cfdf3221e2,On Time,1,Not Repeat
1,0000b849f77a49e4a4ce2b2a4ca5be3f,On Time,1,Not Repeat
2,0000f46a3911fa3c0805444483337064,On Time,1,Not Repeat
3,0000f6ccb0745a6a4b88665a16c9f078,On Time,1,Not Repeat
4,0004aac84e0df4da2b147fca70cf8255,On Time,1,Not Repeat
5,0004bd2a26a76fe21f786e4fbd80607f,On Time,1,Not Repeat
6,00050ab1314c0e55a6ca13cf7181fecf,On Time,1,Not Repeat
7,00053a61a98854899e70ed204dd4bafe,On Time,1,Not Repeat
8,0005e1862207bf6ccc02e4228effd9a0,On Time,1,Not Repeat
9,0005ef4cd20d2893f0d9fbd94d3c0d97,Late,1,Not Repeat


In [40]:
post_delivery_repeat_base_query = """
WITH ranked_delivered_orders AS (

    SELECT
        c.customer_unique_id,
        o.order_id,
        o.order_purchase_timestamp,
        o.order_delivered_customer_date,
        o.order_estimated_delivery_date,

        ROW_NUMBER() OVER (
            PARTITION BY c.customer_unique_id
            ORDER BY o.order_purchase_timestamp
        ) AS order_number

    FROM orders o

    JOIN customers c
        ON o.customer_id = c.customer_id

    WHERE o.order_status = 'delivered'
    AND o.order_delivered_customer_date IS NOT NULL
    AND o.order_estimated_delivery_date IS NOT NULL
),

first_deliveries AS (

    SELECT
        customer_unique_id,
        order_id,
        order_delivered_customer_date,

        CASE
            WHEN order_delivered_customer_date
                 > order_estimated_delivery_date
            THEN 'Late'
            ELSE 'On Time'
        END AS first_delivery_experience

    FROM ranked_delivered_orders

    WHERE order_number = 1
)

SELECT
    f.customer_unique_id,
    f.first_delivery_experience,
    f.order_delivered_customer_date,

    COUNT(o.order_id) AS orders_after_delivery,

    CASE
        WHEN COUNT(o.order_id) > 0
        THEN 'Repeat After Delivery'
        ELSE 'No Repeat After Delivery'
    END AS repeat_status

FROM first_deliveries f

LEFT JOIN customers c
    ON f.customer_unique_id = c.customer_unique_id

LEFT JOIN orders o
    ON c.customer_id = o.customer_id
    AND o.order_purchase_timestamp > f.order_delivered_customer_date

GROUP BY
    f.customer_unique_id,
    f.first_delivery_experience,
    f.order_delivered_customer_date

ORDER BY f.customer_unique_id;
"""

post_delivery_repeat_base = pd.read_sql_query(
    post_delivery_repeat_base_query,
    conn
)

post_delivery_repeat_base.head(10)

,customer_unique_id,first_delivery_experience,order_delivered_customer_date,orders_after_delivery,repeat_status
0,0000366f3b9a7992bf8c76cfdf3221e2,On Time,2018-05-16 20:48:37,0,No Repeat After Delivery
1,0000b849f77a49e4a4ce2b2a4ca5be3f,On Time,2018-05-10 18:02:42,0,No Repeat After Delivery
2,0000f46a3911fa3c0805444483337064,On Time,2017-04-05 14:38:47,0,No Repeat After Delivery
3,0000f6ccb0745a6a4b88665a16c9f078,On Time,2017-11-01 21:23:05,0,No Repeat After Delivery
4,0004aac84e0df4da2b147fca70cf8255,On Time,2017-11-27 23:08:56,0,No Repeat After Delivery
5,0004bd2a26a76fe21f786e4fbd80607f,On Time,2018-04-07 16:12:43,0,No Repeat After Delivery
6,00050ab1314c0e55a6ca13cf7181fecf,On Time,2018-04-27 12:08:59,0,No Repeat After Delivery
7,00053a61a98854899e70ed204dd4bafe,On Time,2018-03-16 19:02:51,0,No Repeat After Delivery
8,0005e1862207bf6ccc02e4228effd9a0,On Time,2017-03-09 08:33:08,0,No Repeat After Delivery
9,0005ef4cd20d2893f0d9fbd94d3c0d97,Late,2018-05-05 12:38:38,0,No Repeat After Delivery


In [41]:
post_delivery_repeat_analysis_query = """
WITH ranked_delivered_orders AS (

    SELECT
        c.customer_unique_id,
        o.order_id,
        o.order_purchase_timestamp,
        o.order_delivered_customer_date,
        o.order_estimated_delivery_date,

        ROW_NUMBER() OVER (
            PARTITION BY c.customer_unique_id
            ORDER BY o.order_purchase_timestamp
        ) AS order_number

    FROM orders o

    JOIN customers c
        ON o.customer_id = c.customer_id

    WHERE o.order_status = 'delivered'
    AND o.order_delivered_customer_date IS NOT NULL
    AND o.order_estimated_delivery_date IS NOT NULL
),

first_deliveries AS (

    SELECT
        customer_unique_id,
        order_delivered_customer_date,

        CASE
            WHEN order_delivered_customer_date
                 > order_estimated_delivery_date
            THEN 'Late'
            ELSE 'On Time'
        END AS first_delivery_experience

    FROM ranked_delivered_orders

    WHERE order_number = 1
),

customer_repeat_behavior AS (

    SELECT
        f.customer_unique_id,
        f.first_delivery_experience,

        CASE
            WHEN COUNT(o.order_id) > 0
            THEN 'Repeat After Delivery'
            ELSE 'No Repeat After Delivery'
        END AS repeat_status

    FROM first_deliveries f

    LEFT JOIN customers c
        ON f.customer_unique_id = c.customer_unique_id

    LEFT JOIN orders o
        ON c.customer_id = o.customer_id
        AND o.order_purchase_timestamp > f.order_delivered_customer_date

    GROUP BY
        f.customer_unique_id,
        f.first_delivery_experience
)

SELECT
    first_delivery_experience,

    COUNT(*) AS total_customers,

    SUM(
        CASE
            WHEN repeat_status = 'Repeat After Delivery'
            THEN 1
            ELSE 0
        END
    ) AS repeat_customers,

    ROUND(
        100.0 *
        SUM(
            CASE
                WHEN repeat_status = 'Repeat After Delivery'
                THEN 1
                ELSE 0
            END
        ) / COUNT(*),
        2
    ) AS repeat_rate

FROM customer_repeat_behavior

GROUP BY first_delivery_experience

ORDER BY first_delivery_experience;
"""

post_delivery_repeat_analysis = pd.read_sql_query(
    post_delivery_repeat_analysis_query,
    conn
)

post_delivery_repeat_analysis

,first_delivery_experience,total_customers,repeat_customers,repeat_rate
0,Late,7604,97,1.28
1,On Time,85746,1747,2.04


In [42]:
delivery_time_by_state_query = """
SELECT
    c.customer_state,

    COUNT(o.order_id) AS delivered_orders,

    ROUND(
        AVG(
            julianday(o.order_delivered_customer_date)
            -
            julianday(o.order_purchase_timestamp)
        ),
        2
    ) AS average_delivery_days,

    ROUND(
        MIN(
            julianday(o.order_delivered_customer_date)
            -
            julianday(o.order_purchase_timestamp)
        ),
        2
    ) AS minimum_delivery_days,

    ROUND(
        MAX(
            julianday(o.order_delivered_customer_date)
            -
            julianday(o.order_purchase_timestamp)
        ),
        2
    ) AS maximum_delivery_days

FROM orders o

JOIN customers c
    ON o.customer_id = c.customer_id

WHERE o.order_status = 'delivered'
AND o.order_purchase_timestamp IS NOT NULL
AND o.order_delivered_customer_date IS NOT NULL

GROUP BY c.customer_state

ORDER BY average_delivery_days DESC;
"""

delivery_time_by_state = pd.read_sql_query(
    delivery_time_by_state_query,
    conn
)

delivery_time_by_state

,customer_state,delivered_orders,average_delivery_days,minimum_delivery_days,maximum_delivery_days
0,RR,41,29.39,6.74,172.14
1,AP,67,27.19,5.11,187.20
2,AM,145,26.43,4.12,138.88
3,AL,397,24.54,4.19,90.96
4,PA,946,23.77,4.38,195.63
5,MA,717,21.57,3.08,168.08
6,SE,335,21.52,6.14,194.63
7,CE,1279,21.27,2.85,168.18
8,AC,80,21.04,7.18,72.20
9,PB,517,20.43,5.73,102.17


In [43]:
estimated_vs_actual_query = """
SELECT

    COUNT(*) AS valid_delivered_orders,

    ROUND(
        AVG(
            julianday(order_delivered_customer_date)
            -
            julianday(order_estimated_delivery_date)
        ),
        2
    ) AS average_days_difference,

    ROUND(
        MIN(
            julianday(order_delivered_customer_date)
            -
            julianday(order_estimated_delivery_date)
        ),
        2
    ) AS earliest_delivery_difference,

    ROUND(
        MAX(
            julianday(order_delivered_customer_date)
            -
            julianday(order_estimated_delivery_date)
        ),
        2
    ) AS latest_delivery_difference,

    SUM(
        CASE
            WHEN order_delivered_customer_date
                 <= order_estimated_delivery_date
            THEN 1
            ELSE 0
        END
    ) AS on_time_or_early_orders,

    SUM(
        CASE
            WHEN order_delivered_customer_date
                 > order_estimated_delivery_date
            THEN 1
            ELSE 0
        END
    ) AS late_orders

FROM orders

WHERE order_status = 'delivered'
AND order_delivered_customer_date IS NOT NULL
AND order_estimated_delivery_date IS NOT NULL;
"""

estimated_vs_actual = pd.read_sql_query(
    estimated_vs_actual_query,
    conn
)

estimated_vs_actual

,valid_delivered_orders,average_days_difference,earliest_delivery_difference,latest_delivery_difference,on_time_or_early_orders,late_orders
0,96470,-11.18,-146.02,188.98,88644,7826


# 🚚 DELIVERY PERFORMANCE ANALYSIS

## 17.1 Order Status Distribution

### Business Question
What percentage of orders were delivered, cancelled, or in other stages?

### Key Concepts
- COUNT()
- GROUP BY
- Percentage calculation

### Result
- Delivered: 96,478 (97.02%)
- Shipped: 1,107 (1.11%)
- Cancelled: 625 (0.63%)
- Unavailable: 609 (0.61%)
- Other statuses made up a very small percentage.

### Insight
The overwhelming majority of orders were successfully delivered.

---

# 17.2 Average Delivery Time

### Business Question
How long does it take for an order to reach the customer?

### Formula

Delivery Time =
Order Delivered Customer Date
-
Order Purchase Timestamp

### Query

SELECT
    COUNT(*) AS delivered_orders,

    ROUND(
        AVG(
            julianday(order_delivered_customer_date)
            -
            julianday(order_purchase_timestamp)
        ),
        2
    ) AS average_delivery_days,

    ROUND(
        MIN(
            julianday(order_delivered_customer_date)
            -
            julianday(order_purchase_timestamp)
        ),
        2
    ) AS minimum_delivery_days,

    ROUND(
        MAX(
            julianday(order_delivered_customer_date)
            -
            julianday(order_purchase_timestamp)
        ),
        2
    ) AS maximum_delivery_days

FROM orders

WHERE order_status = 'delivered'
AND order_purchase_timestamp IS NOT NULL
AND order_delivered_customer_date IS NOT NULL;

### Result
- Delivered orders: 96,470 valid records
- Average delivery time: 12.56 days
- Minimum: 0.53 days
- Maximum: 209.63 days

### Important Data Validation

There were:
- 96,478 delivered orders
- 8 missing delivery dates
- 96,470 valid delivery-time records

Always validate NULL values before calculating date differences.

---

# 17.3 Late Delivery Analysis

### Business Question
How many delivered orders arrived after the estimated delivery date?

### Logic

Late Delivery:

Actual Delivery Date
>
Estimated Delivery Date

### Query Concept

CASE
    WHEN order_delivered_customer_date
         > order_estimated_delivery_date
    THEN 1
    ELSE 0
END

### Result
- Valid delivered orders: 96,470
- Late orders: 7,826
- On-time or early orders: 88,644
- Late delivery rate: 8.11%

### Formula

Late Delivery Rate =
Late Orders / Valid Delivered Orders × 100

### Insight
Approximately 91.89% of valid delivered orders arrived on or before the estimated delivery date.

---

# Late Delivery Rate by State

### Business Question
Which customer states experience the highest late-delivery rates?

### Concepts
- JOIN
- GROUP BY
- CASE WHEN
- Percentage calculation

### Important Result
Some states had significantly higher late-delivery rates.

Examples:
- AL: 23.93%
- MA: 19.67%
- PI: 15.97%
- CE: 15.32%

SP had:
- 5.89% late delivery rate

### Important Insight
High late-delivery rates indicate that actual deliveries exceeded estimated delivery dates more frequently.

However:
Late Delivery Rate ≠ Average Delivery Time.

A state can have long delivery times but still have a low late-delivery rate if the estimated delivery date is sufficiently long.

# 🚚 DELIVERY PERFORMANCE ANALYSIS — PART 2

## 17.4 First Delivery Experience vs Future Repeat Purchase

### Business Question
Are customers who experience a late first delivery less likely to purchase again AFTER receiving that order?

### Why This Analysis Was Improved

A customer might place a second order before receiving their first order.

Example:

Order 1 placed → Jan 1
Order 2 placed → Jan 5
Order 1 delivered late → Jan 20

Therefore, simply checking:

Total Orders > 1

does NOT prove that the customer repeated AFTER experiencing the delivery.

The improved analysis checks:

Another Order Purchase Timestamp
>
First Order Delivery Date

---

### Key SQL Concepts

- CTE (WITH)
- ROW_NUMBER()
- PARTITION BY
- LEFT JOIN
- CASE WHEN
- Date comparison
- GROUP BY

---

### Logic

Step 1:
Find each customer's first valid delivered order.

Step 2:
Classify the first delivery:

Late:
Actual Delivery Date > Estimated Delivery Date

On Time:
Actual Delivery Date <= Estimated Delivery Date

Step 3:
Check whether the customer placed another order AFTER the first order was delivered.

---

### Why LEFT JOIN?

LEFT JOIN keeps all customers, including customers who never placed another order.

If INNER JOIN was used:
Customers with no later orders would disappear.

That would produce an incorrect repeat rate.

---

### Results

Late First Delivery:
- Total customers: 7,604
- Repeat customers after delivery: 97
- Repeat rate: 1.28%

On-Time First Delivery:
- Total customers: 85,746
- Repeat customers after delivery: 1,747
- Repeat rate: 2.04%

### Insight

Customers with an on-time first delivery had a higher post-delivery repeat rate than customers with a late first delivery.

IMPORTANT:

This shows an association.

Do NOT say:
"Late delivery caused customers not to return."

Correct interpretation:

"Late first deliveries were associated with a lower post-delivery repeat rate."

---

# 17.5 Average Delivery Time by State

### Business Question

Which customer states experience the longest and shortest delivery times?

### Query Concepts

- JOIN
- GROUP BY
- AVG()
- MIN()
- MAX()
- julianday()

### Formula

Average Delivery Days =
AVG(
    Delivery Date
    -
    Purchase Date
)

### Results

Longest average delivery times:

- RR: 29.39 days
- AP: 27.19 days
- AM: 26.43 days
- AL: 24.54 days
- PA: 23.77 days

Shortest average delivery times:

- SP: 8.76 days
- PR: 11.99 days
- MG: 12.01 days
- DF: 12.97 days

### Important Insight

Delivery times vary substantially across customer states.

Important:

States with low order volumes should be interpreted cautiously.

Example:

RR:
- 41 delivered orders
- Average: 29.39 days

SP:
- 40,494 delivered orders
- Average: 8.76 days

A small number of unusual orders can influence averages in low-volume states.

---

# 17.6 Estimated vs Actual Delivery Performance

### Business Question

How early or late are orders delivered compared to the estimated delivery date?

### Formula

Delivery Difference =

Actual Delivery Date
-
Estimated Delivery Date

### Interpretation

Negative → Delivered Early

Zero → Delivered exactly on the estimated date

Positive → Delivered Late

---

### Query Concepts

- AVG()
- MIN()
- MAX()
- CASE WHEN
- julianday()

---

### Results

- Valid delivered orders: 96,470
- Average difference: -11.18 days
- Earliest difference: -146.02 days
- Latest difference: 188.98 days
- On-time or early orders: 88,644
- Late orders: 7,826

### Interpretation

Average difference = -11.18 days.

This means:

On average, orders were delivered approximately
11.18 days BEFORE the estimated delivery date.

### Important Insight

Estimated delivery dates appear to have been relatively conservative.

Possible explanations:

- Large delivery buffers
- Conservative delivery estimates
- Deliveries often completed earlier than promised

We cannot determine the exact reason from this dataset alone.

---

# FINAL DELIVERY PERFORMANCE INSIGHTS

1. 97.02% of orders were delivered.

2. Average delivery time was approximately 12.56 days.

3. 8.11% of valid delivered orders were late.

4. 91.89% of valid delivered orders were delivered on time or early.

5. Delivery duration varies substantially across geographic states.

6. Orders were delivered approximately 11.18 days before the estimated delivery date on average.

7. Customers whose first delivery was on time had a higher post-delivery repeat rate (2.04%) than customers whose first delivery was late (1.28%).

IMPORTANT:
The repeat-purchase finding shows association, NOT causation.

In [44]:
review_distribution_query = """
SELECT
    review_score,

    COUNT(*) AS total_reviews,

    ROUND(
        COUNT(*) * 100.0 /
        SUM(COUNT(*)) OVER (),
        2
    ) AS percentage_of_reviews

FROM reviews

GROUP BY review_score

ORDER BY review_score;
"""

review_distribution = pd.read_sql_query(
    review_distribution_query,
    conn
)

review_distribution

,review_score,total_reviews,percentage_of_reviews
0,1,11424,11.51
1,2,3151,3.18
2,3,8179,8.24
3,4,19142,19.29
4,5,57328,57.78


In [45]:
delivery_review_query = """
SELECT
    CASE
        WHEN o.order_delivered_customer_date
             > o.order_estimated_delivery_date
        THEN 'Late'

        ELSE 'On Time or Early'
    END AS delivery_status,

    COUNT(DISTINCT o.order_id) AS total_orders,

    COUNT(r.review_id) AS total_reviews,

    ROUND(AVG(r.review_score), 2) AS average_review_score

FROM orders o

JOIN reviews r
    ON o.order_id = r.order_id

WHERE o.order_status = 'delivered'
AND o.order_delivered_customer_date IS NOT NULL
AND o.order_estimated_delivery_date IS NOT NULL
AND r.review_score IS NOT NULL

GROUP BY delivery_status

ORDER BY average_review_score DESC;
"""

delivery_review = pd.read_sql_query(
    delivery_review_query,
    conn
)

delivery_review

,delivery_status,total_orders,total_reviews,average_review_score
0,On Time or Early,88163,88653,4.29
1,Late,7661,7700,2.57


In [46]:
review_coverage_query = """
SELECT
    COUNT(DISTINCT o.order_id) AS valid_delivered_orders,

    COUNT(DISTINCT r.order_id) AS orders_with_reviews,

    COUNT(DISTINCT o.order_id)
    -
    COUNT(DISTINCT r.order_id) AS orders_without_reviews

FROM orders o

LEFT JOIN reviews r
    ON o.order_id = r.order_id

WHERE o.order_status = 'delivered'
AND o.order_delivered_customer_date IS NOT NULL
AND o.order_estimated_delivery_date IS NOT NULL;
"""

review_coverage = pd.read_sql_query(
    review_coverage_query,
    conn
)

review_coverage

,valid_delivered_orders,orders_with_reviews,orders_without_reviews
0,96470,95824,646


# ⭐ CUSTOMER SATISFACTION ANALYSIS

## 18.1 Review Score Distribution

### Business Question
How satisfied are customers overall?

### Concepts Used
- GROUP BY
- COUNT()
- Percentage calculation
- Window function: OVER()

### Results

| Review Score | Total Reviews | Percentage |
|---|---:|---:|
| 1 Star | 11,424 | 11.51% |
| 2 Stars | 3,151 | 3.18% |
| 3 Stars | 8,179 | 8.24% |
| 4 Stars | 19,142 | 19.29% |
| 5 Stars | 57,328 | 57.78% |

### Key Findings

- The most common rating was 5 stars.
- 57.78% of all reviews were 5-star reviews.
- 77.07% of reviews were either 4 or 5 stars.
- 14.69% of reviews were either 1 or 2 stars.

### Insight

Overall customer satisfaction appears high, with more than three-quarters of reviews receiving a rating of 4 or 5 stars.

However, 11.51% of reviews were 1-star ratings, indicating that a meaningful portion of customers had a poor experience.

---

# 18.2 Delivery Performance vs Customer Reviews

### Business Question

Are late deliveries associated with lower customer satisfaction?

### Tables Used

orders
reviews

### Join

orders.order_id = reviews.order_id

### Delivery Classification

Late:

Actual Delivery Date
>
Estimated Delivery Date

On Time or Early:

Actual Delivery Date
<=
Estimated Delivery Date

### Concepts Used

- JOIN
- CASE WHEN
- COUNT()
- COUNT(DISTINCT)
- AVG()
- GROUP BY

### Results

| Delivery Status | Total Orders | Total Reviews | Average Review Score |
|---|---:|---:|---:|
| On Time or Early | 88,163 | 88,653 | 4.29 |
| Late | 7,661 | 7,700 | 2.57 |

### Key Finding

On-time or early deliveries had an average review score of 4.29.

Late deliveries had an average review score of 2.57.

Difference:

4.29 - 2.57 = 1.72 stars

### Insight

Orders delivered late were associated with substantially lower customer satisfaction compared with orders delivered on time or early.

IMPORTANT:

This analysis shows association, NOT causation.

Correct wording:

"Late deliveries were associated with lower review scores."

Do NOT say:

"Late deliveries caused lower customer ratings."

Other factors may also influence review scores.

---

# Data Validation — Review Coverage

Before interpreting the results, review coverage was checked.

| Metric | Count |
|---|---:|
| Valid delivered orders | 96,470 |
| Orders with reviews | 95,824 |
| Orders without reviews | 646 |

Validation:

95,824 + 646 = 96,470

Therefore, 646 valid delivered orders did not have a matching review.

The delivery vs review analysis includes only orders with available review data.

---

# FINAL CUSTOMER SATISFACTION INSIGHTS

1. Overall customer satisfaction was generally high.

2. 57.78% of all reviews received a 5-star rating.

3. 77.07% of reviews received either 4 or 5 stars.

4. Late deliveries were strongly associated with lower customer satisfaction.

5. On-time or early deliveries had an average review score of 4.29.

6. Late deliveries had an average review score of 2.57.

7. The difference between the two groups was 1.72 stars.

8. This relationship represents association, not proven causation.

In [47]:
top_sellers_revenue_query = """
SELECT
    s.seller_id,
    s.seller_city,
    s.seller_state,

    ROUND(SUM(oi.price), 2) AS product_revenue,

    COUNT(DISTINCT oi.order_id) AS total_orders,

    COUNT(*) AS total_items_sold

FROM order_items oi

JOIN sellers s
    ON oi.seller_id = s.seller_id

GROUP BY
    s.seller_id,
    s.seller_city,
    s.seller_state

ORDER BY product_revenue DESC

LIMIT 15;
"""

top_sellers_revenue = pd.read_sql_query(
    top_sellers_revenue_query,
    conn
)

top_sellers_revenue

,seller_id,seller_city,seller_state,product_revenue,total_orders,total_items_sold
0,4869f7a5dfa277a7dca6462dcf3b52b2,guariba,SP,229472.63,1132,1156
1,53243585a1d6dc2643021fd1853d8905,lauro de freitas,BA,222776.05,358,410
2,4a3ca9315b744ce9f8e9374361493884,ibitinga,SP,200472.92,1806,1987
3,fa1c13f2614d7b5c4749cbc52fecda94,sumare,SP,194042.03,585,586
4,7c67e1448b00f6e969d365cea6b010ab,itaquaquecetuba,SP,187923.89,982,1364
5,7e93a43ef30c4f03f38b393420bc753a,barueri,SP,176431.87,336,340
6,da8622b14eb17ae2831f4ac5b9dab84a,piracicaba,SP,160236.57,1314,1551
7,7a67c85e85bb2ce8582c35f2203ad736,sao paulo,SP,141745.53,1160,1171
8,1025f0e2d44d7041d6cf58b6550e0bfa,sao paulo,SP,138968.55,915,1428
9,955fee9216a65b617aa5c0531780ce60,sao paulo,SP,135171.70,1287,1499


In [48]:
seller_value_query = """
SELECT
    s.seller_id,
    s.seller_city,
    s.seller_state,

    ROUND(SUM(oi.price), 2) AS product_revenue,

    COUNT(DISTINCT oi.order_id) AS total_orders,

    COUNT(*) AS total_items_sold,

    ROUND(
        SUM(oi.price) * 1.0 / COUNT(*),
        2
    ) AS average_revenue_per_item

FROM order_items oi

JOIN sellers s
    ON oi.seller_id = s.seller_id

GROUP BY
    s.seller_id,
    s.seller_city,
    s.seller_state

ORDER BY product_revenue DESC

LIMIT 15;
"""

seller_value = pd.read_sql_query(
    seller_value_query,
    conn
)

seller_value

,seller_id,seller_city,seller_state,product_revenue,total_orders,total_items_sold,average_revenue_per_item
0,4869f7a5dfa277a7dca6462dcf3b52b2,guariba,SP,229472.63,1132,1156,198.51
1,53243585a1d6dc2643021fd1853d8905,lauro de freitas,BA,222776.05,358,410,543.36
2,4a3ca9315b744ce9f8e9374361493884,ibitinga,SP,200472.92,1806,1987,100.89
3,fa1c13f2614d7b5c4749cbc52fecda94,sumare,SP,194042.03,585,586,331.13
4,7c67e1448b00f6e969d365cea6b010ab,itaquaquecetuba,SP,187923.89,982,1364,137.77
5,7e93a43ef30c4f03f38b393420bc753a,barueri,SP,176431.87,336,340,518.92
6,da8622b14eb17ae2831f4ac5b9dab84a,piracicaba,SP,160236.57,1314,1551,103.31
7,7a67c85e85bb2ce8582c35f2203ad736,sao paulo,SP,141745.53,1160,1171,121.05
8,1025f0e2d44d7041d6cf58b6550e0bfa,sao paulo,SP,138968.55,915,1428,97.32
9,955fee9216a65b617aa5c0531780ce60,sao paulo,SP,135171.70,1287,1499,90.17


# 📦 SELLER PERFORMANCE ANALYSIS

## 19.1 Top Sellers by Product Revenue

### Business Question

Which sellers generate the highest product revenue?

### Tables Used

order_items
sellers

### Join

order_items.seller_id = sellers.seller_id

### Revenue Definition

Seller Product Revenue =

SUM(order_items.price)

Freight value is not included.

This maintains consistency with the earlier product category revenue analysis.

### Concepts Used

- JOIN
- SUM()
- COUNT(DISTINCT)
- COUNT()
- GROUP BY
- ORDER BY
- LIMIT

### Key Results

The highest-revenue seller:

- Location: Guariba, SP
- Product Revenue: 229,472.63
- Orders: 1,132
- Items Sold: 1,156

Other high-revenue sellers included sellers located in:

- Lauro de Freitas, BA
- Ibitinga, SP
- Sumare, SP
- Itaquaquecetuba, SP

### Key Insight

High seller revenue does not necessarily mean the seller has the highest number of orders or items sold.

Revenue and sales volume measure different aspects of seller performance.

---

## 19.2 Seller Revenue vs Sales Volume

### Business Question

Do sellers generate revenue through high-value products or high sales volume?

### Formula

Average Revenue per Item =

Product Revenue
/
Total Items Sold

### Concepts Used

- JOIN
- SUM()
- COUNT()
- COUNT(DISTINCT)
- ROUND()
- GROUP BY

### Key Results

High-value sales examples:

Lauro de Freitas, BA:

- Product Revenue: 222,776.05
- Items Sold: 410
- Average Revenue per Item: 543.36

Barueri, SP:

- Product Revenue: 176,431.87
- Items Sold: 340
- Average Revenue per Item: 518.92

High-volume sales examples:

Ibitinga, SP:

- Product Revenue: 200,472.92
- Items Sold: 1,987
- Average Revenue per Item: 100.89

São Paulo seller:

- Product Revenue: 123,304.83
- Items Sold: 2,033
- Average Revenue per Item: 60.65

### Key Insight

Seller performance follows different patterns.

Some sellers generate high revenue by selling fewer, higher-value items.

Other sellers generate revenue through high sales volume with lower average item values.

Therefore:

High Revenue ≠ High Sales Volume

Both revenue and sales volume should be considered when evaluating seller performance.

---

# FINAL SELLER PERFORMANCE INSIGHTS

1. Sellers differ significantly in revenue generation.

2. The highest-revenue seller generated 229,472.63 in product revenue.

3. High revenue does not necessarily correspond to high sales volume.

4. Some sellers generated substantial revenue from relatively few items, indicating higher average item values.

5. Other sellers generated revenue through large sales volumes with lower average revenue per item.

6. Seller performance should be evaluated using multiple metrics:

- Product Revenue
- Total Orders
- Total Items Sold
- Average Revenue per Item

In [49]:
average_price_category_query = """
SELECT
    p.product_category_name,

    COUNT(*) AS total_items_sold,

    ROUND(AVG(oi.price), 2) AS average_item_price,

    ROUND(SUM(oi.price), 2) AS product_revenue

FROM order_items oi

JOIN products p
    ON oi.product_id = p.product_id

WHERE p.product_category_name IS NOT NULL

GROUP BY p.product_category_name

ORDER BY average_item_price DESC;
"""

average_price_category = pd.read_sql_query(
    average_price_category_query,
    conn
)

average_price_category

,product_category_name,total_items_sold,average_item_price,product_revenue
0,pcs,203,1098.34,222963.13
1,portateis_casa_forno_e_cafe,76,624.29,47445.71
2,eletrodomesticos_2,238,476.12,113317.74
3,agro_industria_e_comercio,212,342.12,72530.47
4,instrumentos_musicais,680,281.62,191498.88
...,...,...,...,...
68,alimentos_bebidas,278,54.60,15179.48
69,cds_dvds_musicais,14,52.14,730.00
70,fraldas_higiene,39,40.19,1567.59
71,flores,33,33.64,1110.04


# 📦 PRODUCT ANALYSIS

## Product Category Sales and Revenue Analysis

### Business Questions

Which product categories have the highest sales volume?

Which product categories generate the highest product revenue?

### Metrics Used

- Total Orders
- Total Items Sold
- Product Revenue

### Revenue Definition

Product Revenue =

SUM(order_items.price)

Freight value was not included in product revenue.

### Key Results

The leading product categories by sales volume included:

- bed_bath_table
- health_beauty
- sports_leisure
- computers_accessories
- furniture_decor

High-revenue categories included:

- health_beauty: 1,233,131.72
- watches_gifts: 1,166,176.98
- bed_bath_table: 1,023,434.76
- sports_leisure: 954,852.55
- computers_accessories: 888,724.61

### Key Insight

High sales volume does not always result in the highest product revenue.

For example, product categories can generate high revenue because of:

- High sales volume
- Higher-priced products
- A combination of both

Therefore, sales volume and revenue should be analyzed together.

---

## Average Item Price by Product Category

### Business Question

Which product categories tend to have higher-priced items?

### Tables Used

- order_items
- products

### Join

order_items.product_id = products.product_id

### Formula

Average Item Price =

AVG(order_items.price)

This represents the average selling price of order items within each product category.

### Key Results

The categories with the highest average item prices included:

- pcs: 1,098.34
- portateis_casa_forno_e_cafe: 624.29
- eletrodomesticos_2: 476.12
- agro_industria_e_comercio: 342.12
- instrumentos_musicais: 281.62

The category `pcs` had:

- 203 items sold
- Average item price: 1,098.34
- Product revenue: 222,963.13

Lower-priced categories included:

- flores: 33.64
- fraldas_higiene: 40.19
- cds_dvds_musicais: 52.14

### Key Insight

Product categories follow different sales patterns.

Some categories generate revenue through high sales volume.

Other categories generate substantial revenue despite lower sales volume because their products have higher average prices.

Therefore:

High Sales Volume ≠ High Revenue

High Revenue ≠ High Average Item Price

Product performance should be evaluated using:

- Sales Volume
- Product Revenue
- Average Item Price

In [50]:
monthly_orders_query = """
SELECT
    strftime('%Y-%m', order_purchase_timestamp) AS order_month,
    
    COUNT(DISTINCT order_id) AS total_orders

FROM orders

WHERE order_purchase_timestamp IS NOT NULL

GROUP BY
    strftime('%Y-%m', order_purchase_timestamp)

ORDER BY order_month;
"""

monthly_orders = pd.read_sql_query(
    monthly_orders_query,
    conn
)

monthly_orders

,order_month,total_orders
0,2016-09,4
1,2016-10,324
2,2016-12,1
3,2017-01,800
4,2017-02,1780
5,2017-03,2682
6,2017-04,2404
7,2017-05,3700
8,2017-06,3245
9,2017-07,4026


In [51]:
monthly_revenue_query = """
SELECT
    strftime('%Y-%m', o.order_purchase_timestamp) AS order_month,

    ROUND(SUM(oi.price), 2) AS product_revenue,

    COUNT(DISTINCT o.order_id) AS total_orders

FROM orders o

JOIN order_items oi
    ON o.order_id = oi.order_id

WHERE o.order_purchase_timestamp IS NOT NULL

GROUP BY
    strftime('%Y-%m', o.order_purchase_timestamp)

ORDER BY order_month;
"""

monthly_revenue = pd.read_sql_query(
    monthly_revenue_query,
    conn
)

monthly_revenue

,order_month,product_revenue,total_orders
0,2016-09,267.36,3
1,2016-10,49507.66,308
2,2016-12,10.90,1
3,2017-01,120312.87,789
4,2017-02,247303.02,1733
5,2017-03,374344.30,2641
6,2017-04,359927.23,2391
7,2017-05,506071.14,3660
8,2017-06,433038.60,3217
9,2017-07,498031.48,3969


# 📅 TIME & TREND ANALYSIS

## Business Objective

Analyze how customer purchasing activity and product revenue changed over time.

This analysis examines:

- Monthly order volume
- Monthly product revenue
- Overall business growth trends

---

# 20.1 Monthly Order Trend

## Business Question

How did customer order volume change over time?

## Table Used

- orders

## Date Used

order_purchase_timestamp

This timestamp represents when the customer placed an order.

## Concepts Used

- strftime()
- COUNT(DISTINCT)
- GROUP BY
- ORDER BY

## Method

Orders were grouped by month using:

strftime('%Y-%m', order_purchase_timestamp)

The number of unique orders was calculated for each month.

## Key Results

Order volume increased substantially throughout 2017.

Examples:

- January 2017: 800 orders
- March 2017: 2,682 orders
- May 2017: 3,700 orders
- August 2017: 4,331 orders
- October 2017: 4,631 orders
- November 2017: 7,544 orders

November 2017 recorded the highest monthly order volume with:

7,544 orders

Order activity remained consistently high throughout most of 2018, generally ranging between approximately 6,000 and 7,200 orders per month.

## Important Data Consideration

The first and last months of the dataset contain unusually low order counts.

Examples:

- September 2016: 4 orders
- October 2016: 324 orders
- September 2018: 16 orders
- October 2018: 4 orders

These periods should not be interpreted as normal business performance because they may represent partial or incomplete dataset coverage.

## Key Insight

Customer purchasing activity grew substantially during 2017 and remained at a relatively high and stable level throughout most of 2018.

---

# 20.2 Monthly Product Revenue Trend

## Business Question

How did product revenue change over time?

## Tables Used

- orders
- order_items

## Join

orders.order_id = order_items.order_id

## Date Used

order_purchase_timestamp

The purchase timestamp was used to assign product revenue to the month in which the customer placed the order.

## Revenue Definition

Product Revenue =

SUM(order_items.price)

Freight value was not included.

## Concepts Used

- JOIN
- SUM()
- COUNT(DISTINCT)
- strftime()
- GROUP BY
- ORDER BY

## Method

Orders were joined with order items using order_id.

Revenue was grouped according to the month in which the customer placed the order.

## Key Results

Product revenue generally increased from early 2017 into 2018.

Examples:

- January 2017: 120,312.87
- March 2017: 374,344.30
- May 2017: 506,071.14
- August 2017: 573,971.68
- October 2017: 664,219.43
- November 2017: 1,010,271.37

November 2017 recorded the highest monthly product revenue:

1,010,271.37

Revenue remained high throughout most of 2018.

Examples:

- January 2018: 950,030.36
- March 2018: 983,213.44
- April 2018: 996,647.75
- May 2018: 996,517.68

## Important Data Consideration

The revenue analysis uses an INNER JOIN between orders and order_items.

Therefore, only orders with matching order item records are included in the revenue calculation.

The total_orders value in this analysis represents orders associated with order items and may differ from the total order count in the Monthly Order Trend analysis.

Very low values in the first and last dataset months should be treated as potentially incomplete periods rather than normal business performance.

## Key Insight

November 2017 was a major peak in both:

- Customer order volume
- Product revenue

Order volume and revenue generally followed similar growth patterns but were not perfectly proportional.

This indicates that revenue depends not only on the number of orders but also on the value of products purchased.

---

# Overall Time & Trend Insight

The business experienced substantial growth during 2017.

Both order volume and product revenue increased significantly, reaching a major peak in November 2017.

Throughout most of 2018, customer purchasing activity and product revenue remained consistently high.

This suggests that the business reached a higher and relatively stable level of activity after its growth during 2017.